### PR TITLE
Add navigation determinant params

### PR DESCRIPTION
**Pull request by joshuatf**
Part 1 of a 2 part exploration to improve route detection.

This PR allows `queryParams` (up for discussion- `requiredParams`?) to be passed along with each individual item to help determine if this item should be marked active or not.  For example passing `'queryParams' => array( 'my_param' )` as a menu item argument means that the current URL location must match the menu item's `my_param` option.

This is useful since some pages are determined active by the `page` param, while others, like those in WCA, are determined by `page` and `path` and those in WC Settings are determined by `tab`.

The follow-up PR will explore a regex alternative.

### Screenshots
Screen Shot 2020-12-02 at 5 39 05 PM

### Detailed test instructions:

1. Add some items that have possible conflict in params due to a matching score being higher.  For example, you could register a new WCA page:

```php
wc_admin_register_page(
	array(
		'id'     => 'test-determinant-params',
		'title'  => __( 'Test Determinant Params', 'woocommerce-admin' ),
		'path'   => '/test-params&param1=a&param2=b',
		'nav_args' => array()
	)
);
```

2. Navigate to a URL where the match score should be higher for that item, but using a non-matching determinant param (`path` in the case of WCA).  `wp-admin/admin.php?page=wc-admin&param1=a&param2=b`
3. Note that the correct item is matched.

### CODE REVIEW

File: client/navigation/utils.js [hunk 1]
```diff
@@ -82,11 +83,17 @@ export const getMatchScore = ( location, url ) => {
...
-		if ( urlParams[ key ] === locationParams[ key ] ) {
+	for ( const key in itemParams ) {
+		if ( itemParams[ key ] === locationParams[ key ] ) {
 			matchingParamCount++;
+		} else if (
+			queryParams.includes( key ) &&
+			( itemParams[ key ] || locationParams[ key ] )
+		) {
+			// A determinant param was found and not matched.
+			return 0;
```
Comment (psealock) [inline client/navigation/utils.js:94]: Nice 👍

**Review by psealock — COMMENTED** [review]
Good start @joshuatf, this is working well and the case encountered in https://github.com/woocommerce/woocommerce-admin/issues/5710 is resolved.

>up for discussion- requiredParams

I think so. Its more explicit. I think there could be a way for us to calculate these on behalf of extensions. Can we have a list of params that are added automatically, like `page` and the rest are derived from the supplied url? That list of params could be filterable too.

It feels like this would be one less step for extensions

**Comment by joshuatf** [comment]
>Can we have a list of params that are added automatically, like page and the rest are derived from the supplied url? That list of params could be filterable too.

Great idea.  I was trying to avoid this since even though `page` is used to determine pages in core, I didn't want to impose any restrictions.  But if we create a filterable list of default args, then we get the best of both worlds.

**Comment by psealock** [comment]
I took this for a spin to compare the regex approach. I register a page with a url including a query parameter.

```php
function page_content(){
	echo '<div class="wrap"><h2>Testing</h2></div>';
}

add_menu_page( 'My Extension', 'My Extension', 'manage_woocommerce', 'my-extension-slug', 'page_content' );


Menu::add_plugin_item(
	array(
		'id'         => 'my-extension',
		'title'      => __( 'My Extension', 'my-extension' ),
		'capability'    => 'manage_woocommerce',
		'url'        => 'my-extension-slug&tab=new',
		'parent' => 'woocommerce',
		'queryParams' => array( 'page', 'tab' ),
	)
);
```
When the `queryParams` are supplied, this behaves well. Inclusion of the `page`, and `tab` parameters are the only thing required for a correct match, regardless of order. As mentioned, these parameters don't need to be supplied, but can be derived from the `url` parameter plus a filterable list of definites such as `page`. 

This approach is my vote. We get the desired result without relying on extensions to craft difficult regex or maintain our own default regex that supports variable query parameter order.

**Comment by joshuatf** [comment]
Closing this in favor of https://github.com/woocommerce/woocommerce-admin/pull/5818

**Pull request closed without merging**